engines/scumm/saveload.cpp

```diff
@@ -178,6 +175,323 @@ static bool saveSaveGameHeader(Common::WriteStream *out, SaveGameHeader &hdr) {
+			color = remapPaletteColor(
+				brightness * ((rgb & 0xFF) >> 0) / 0xFF,
+				brightness * ((rgb & 0xFF00) >> 8) / 0xFF,
+				brightness * ((rgb & 0xFF0000) >> 16) / 0xFF,
+				-1);
+
+			// The -1 after boxHeight is done to compensate for the fact that
+			// we can't directly control the back and front buffers (see below)
+			drawBox(boxX, boxY, boxWidth, boxHeight - 1, color);
+			return;
+		}
+	}
+
+	// If we got here, it means we managed to fetch one of the
+	// thumbnails, so let's actually draw it to screen!
+	heightSlice = 0;
+	for (int i = 0; i < boxHeight; i++) {
+		pixelY = boxY + i;
+		widthSlice = 0;
+		for (int j = 0; j < boxWidth; j++) {
+			pixelX = j + boxX;
+
+			// Remember, the internal one is paletted, while the ScummVM one
+			// is blitted without going through a palette index...
+			if (foundInternalThumbnail) {
+				color = _savegameThumbnail[160 * (heightSlice / boxHeight) + (widthSlice / boxWidth)];
+				pixelColor = tmpPalette[color];
+			} else {
+				pixelColor = thumbSurface[160 * (heightSlice / boxHeight) + (widthSlice / boxWidth)];
+			}
+
+			// Draw twice; once in the frontbuffer, once in the backbuffer:
+			// this ensures that the lowest row of the image doesn't get overwritten
+			// by the blastText rect just below, containing the savegame name...
+			drawPixel(vs, pixelX, pixelY, pixelColor, false);
+			drawPixel(vs, pixelX, pixelY, pixelColor, true);
+
+			widthSlice += 160;
+		}
+		heightSlice += 120;
+	}
+
+	if (thumbSurface)
+		free(thumbSurface);
+}
+
+void ScummEngine_v8::createInternalSaveStateThumbnail() {
+	byte *tempBitmap = (byte *)malloc(_screenWidth * _screenHeight * sizeof(byte));
+	VirtScreen *vs = &_virtscr[kMainVirtScreen];
+
+
+	byte *screen = vs->getPixels(0, _screenTop);
+
+	if (tempBitmap) {
+		for (int i = 0; i < _screenHeight; i++) {
+			screen = vs->getPixels(0, _screenTop + i);
+			memcpy(&tempBitmap[_screenWidth * i], screen, _screenWidth * sizeof(byte));
+		}
+
+		for (int i = 0; i < 256; i++) {
+			_savegameThumbnailPalette[i] = getPackedRGBColorFromPalette(_currentPalette, i);
+		}
+
+		for (int i = 0; i < 120; i++) {
+			for (int j = 0; j < 160; j++) {
+				_savegameThumbnail[i * 160 + j] = tempBitmap[4 * (i * _screenWidth + j)];
+			}
+		}
+
+		free(tempBitmap);
+	}
+}
+
+bool ScummEngine_v8::fetchInternalSaveStateThumbnail(int slotId, bool isHeapSave) {
+	SaveGameHeader hdr;
+	Common::String filename;
+	Common::SeekableReadStream *in = openSaveFileForReading(slotId, isHeapSave, filename);
+	if (!in)
+		return false;
+
+	// In order to fetch the internal COMI thumbnail, we perform the same routine
+	// used during normal loading, stripped down to support only version 106 onwards...
+	if (!loadAndCheckSaveGameHeader(in, _game.heversion, hdr)) {
+		delete in;
+		return false;
+	}
+
+	if (hdr.ver > 0xFFFFFF)
+		hdr.ver = SWAP_BYTES_32(hdr.ver);
+
+	// Reject save games which do not contain the internal thumbnail...
+	if (hdr.ver < VER(106)) {
+		return false;
```

Comment:
you miss a `delete in` here.
This must be the reason of your leak.

Reply:
I will test that as soon as I can get my hands on my laptop, but thanks already! 🙏

Reply:
Works like a charm and even fixes the other related leak in `fetchScummVMSaveStateThumbnail()` somehow! Thanks for that! 🙏